opal/mca/common/ofi/common_ofi.c

```diff
@@ -784,110 +588,190 @@ static uint32_t get_package_rank(opal_process_info_t *process_info)
+                return OPAL_ERROR;
+            }
 
-    /* Initialize opal_hwloc_topology if it is not already */
-    ret = opal_hwloc_base_get_topology();
-    if (0 > ret) {
-        /* Provider selection can continue but there is no guarantee of locality */
-        opal_output_verbose(1, opal_common_ofi.output, "%s:%d:Failed to initialize topology\n",
+            common_ancestor = hwloc_get_common_ancestor_obj(opal_hwloc_topology, ojb, prov_dev);
+            if (!common_ancestor) {
+                opal_output_verbose(1, opal_common_ofi.output,
+                                    "%s:%d:Failed to find common ancestor with provider PCI device",
+                                    __FILE__, __LINE__);
+                /* Any 2 objects should share at least one common ancestor, i.e. root */
+                return OPAL_ERROR;
+            }
+
+            ret = get_obj_depth(common_ancestor, &depth);
+            if (OPAL_SUCCESS != ret) {
+                opal_output_verbose(1, opal_common_ofi.output,
+                                    "%s:%d:Failed to get common ancestor depth", __FILE__,
+                                    __LINE__);
+                return OPAL_ERROR;
+            }
+
+            if (max_common_ancestor_depth < depth) {
+                max_common_ancestor_depth = depth;
+                near_provider_count = 1;
+            } else if (max_common_ancestor_depth == depth) {
+                ++near_provider_count;
+            }
+        }
+
+        *(distance++) = !common_ancestor ? 0 : depth;
+    }
+
+    if (0 == near_provider_count || 0 > max_common_ancestor_depth) {
+        opal_output_verbose(1, opal_common_ofi.output, "%s:%d:Provider does not have PCI device",
                             __FILE__, __LINE__);
+        return OPAL_ERR_NOT_AVAILABLE;
     }
 
-    provider_limit = count_providers(provider_list);
+    provider_rank = rank % near_provider_count;
 
-    /* Allocate memory for provider table */
-    provider_table = calloc(provider_limit, sizeof(struct fi_info *));
-    if (NULL == provider_table) {
-        opal_output_verbose(1, opal_common_ofi.output,
-                            "%s:%d:Failed to allocate memory for provider table\n", __FILE__,
-                            __LINE__);
-        return provider_list;
+    distance = distances;
+    current_provider = provider_list;
+    while (NULL != current_provider) {
+        if ((uint32_t) max_common_ancestor_depth == *(distance++)
+            && provider_rank == --near_provider_count) {
+            *provider = current_provider;
+            return OPAL_SUCCESS;
+        }
+
+        current_provider = current_provider->next;
     }
 
-#if OPAL_OFI_PCI_DATA_AVAILABLE
-    /* find all the nearest devices to this thread, then out of these
-     * determine which device we should bind to.
-     */
-    distances = get_nearest_nics(&num_distances, &pmix_val);
-#endif
+    assert(0 == near_provider_count);
```

Comment:
is this assert necessary in light of the logic at line 655?  we normally try to be very judicious about using asserts as its not a nice way to fail.

Reply:
I'm not sure I agree with @hppritcha here.  We try to avoid assert/abort as an error handler, for sure.  But we have asserts all over the code for things that are internal programming error catchers, which this one clearly is.  If near_provider_count == 0 at this point, the code is wrong.  It's not a user bug or a calling convention bug or anything like that, but that this particular piece of code is wrong.  So I'm happy with this assert.